### PR TITLE
VPN-6422: Migrate to XDG Secrets Portal

### PR DESCRIPTION
**Pull request by oskirby**
## Description
We have a steady stream of users reporting that their Linux clients are frequently loosing their settings across reboot. This looks and smells like a failure in the `LinuxCryptoSettings` class to retrieve the settings encryption keys from the `libsecret` API. We have seen this before, and it typically results from the user's keyring being locked at the time that they launch the VPN client.

Fortunately, there is a better way to generate an application encryption key using the XDG [Secrets Portal](https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.portal.Secret.html), and we already have it implemented for Flatpaks. This PR attempts to provide a migration strategy for users who have settings encrypted with `libsecret` to switch over to the XDG secrets instead.

We accomplish this by doing the following:
1. Extend the `CryptoSettings::getKey()` method to pass in the encrypted settings file version. This lets us distinguish if a file was encrypted with `libsecret` or the XDG portal.
2. Replace `LinuxCryptoSettings` with a wrapper that checks for a key in `libsecret` to use as a fallback if a V1 file is found, otherwise it just wraps `XdgCryptoSettings` instead.
3. Add some code to delete the keys in `libsecret` once the user attempts to write a file with an XDG secret.

Along the way, we found a bug (or quirk?). When the application is launched via the old `/etc/xdg/autostart` handlers, this results in the application being unassociated with a systemd scope. This means that any attempt to use an XDG portal results in the portals being unable to figure out the application ID. This can result in the application getting a different set of keys than when launched via the `.desktop` file. Different keys results in a failure to read the settings files, which results in an apparent loss of settings. Luckily for us, this can be addressed by using the XDG [Background Portal](https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.portal.Background.html) for launching at boot... which we also have implemented as a part of the Flatpak work.

There is a potential bug in here though, or maybe a just a developer annoyance. Because the XDG secrets portal is sensitive to the application ID (which in turn is derived from the systemd scopes it is launched in). We wind up with different encryption keys if we launch it via the application menu (which winds up with the proper scopes and app IDs) vs. when we start it manually on the command line (which gets no scope).

And there seems to be something odd in the RPM build world that breaks when trying to include the Qt6 private GUI libraries (for the XDG Portal helpers). This doesn't seem to cause problems on Fedora 39 or later, so we have also included an update to the Fedora build images here as well (they were long overdue for an update anyways as the old versions have been EOL for a number of months).

## Reference
JIRA Issue: [VPN-6422](https://mozilla-hub.atlassian.net/browse/VPN-6422)
User bug reports:
- #9624
- #6607
- #6150 

## Checklist
    
- [ ] My code follows the style guidelines for this project
- [ ] I have not added any packages that contain high risk or unknown licenses (GPL,  LGPL, MPL, etc. consult with DevOps if in question)
- [ ] I have performed a self review of my own code
- [ ] I have commented my code PARTICULARLY in hard to understand areas
- [ ] I have added thorough tests where needed

[VPN-6422]: https://mozilla-hub.atlassian.net/browse/VPN-6422?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by ahal** [comment]
r+ for the taskcluster changes, though removing myself so this gets a proper review.

File: src/platforms/linux/linuxcryptosettings.h
Comment (mcleinman) [inline src/platforms/linux/linuxcryptosettings.h:8]: Can this line be removed now?

File: src/platforms/linux/linuxcryptosettings.cpp
Comment (mcleinman) [inline src/platforms/linux/linuxcryptosettings.cpp:106]: Should we confirm this returns a key? And if it returns an empty bytearray try to fall back to legacy? Not sure what scenarios, if any, this would help in.

File: src/platforms/linux/linuxcryptosettings.cpp
Comment (mcleinman) [inline src/platforms/linux/linuxcryptosettings.cpp:66]: Do we want to log the error in this block, as we do currently?

File: src/cryptosettings.cpp
Comment (mcleinman) [inline src/cryptosettings.cpp:232]: When do we not have SyrptoSettings? Is it better to fail to plaintext, or fail to not saving the key?

File: src/cryptosettings.cpp
Comment (mcleinman) [inline src/cryptosettings.cpp:335]: We have an early return for ChachaPolyV1. Could this have any bad consequences, especially for non-Linux systems?

**Review by mcleinman — COMMENTED** [review]
A bunch of questions, mostly to make sure I understand this area of the code and the tradeoffs made here.

I'd like this to be a +2 (2 VPN engineers sign off), given the security implications and my limited prior work in this area of the codebase. I'd be comfortable being one of the signoffs after I make sure I'm understanding this - happy to setup time to talk live about my questions, or you're welcome to answer them here - whatever is easier.

File: taskcluster/kinds/build/linux.yml
Comment (oskirby) [reply taskcluster/kinds/build/linux.yml:58]: This PR enables the `XdgPortal` class for all Linux builds (no longer just the Flatpaks), which has a dependency on `Qt6::GuiPrivate` in order to resolve the window handle. It turned out that this dependency doesn't build cleanly on Fedora 37/38.

These versions of Fedora have also been EOL for a couple of months now, so I figured the cleanest solution was just to do an upgrade...  but I could be convinced to ignore the problem and do that upgrade in another PR.

File: taskcluster/kinds/docker-image/kind.yml
Comment (oskirby) [reply taskcluster/kinds/docker-image/kind.yml:60]: See comment above for Fedora 38. TDLR - this introduces a new dependency that fails to build on Fedora 37.

File: src/platforms/linux/xdgportal.cpp
Comment (oskirby) [reply src/platforms/linux/xdgportal.cpp:17]: Now that this file is built on all Linux versions, we found out that it contains a bug on Qt < 6.5.0. In particular, `QGuiApplication` only includes a foreward declaration of `QWindow` but does not define that class. Which means that the call to get the X11 window handle (`holder->window()->winId()`) will fail due to there being on definition of `QWindow::winId()`

File: src/cryptosettings.cpp
Comment (oskirby) [reply src/cryptosettings.cpp:232]: Oops, that was a bit of a typo :)

When `s_instance` is null, this means that there is no platform provider for fetching keys in which case the only format we can possibly support is the plaintext settings (which is what `writeJsonFile()` writes). Nothing has really changed in this method in terms of logic, I just reorganized it to eliminate the local variable.

File: src/cryptosettings.cpp
Comment (oskirby) [reply src/cryptosettings.cpp:335]: Uh, good catch, this if/else flow is bit of a mess.

The intention was that we can `writeEncryptedChachaPolyFile` to write an encrypted settings file then call `getPrefferedKey()` to determine which format the platform wants to use to select V1 or V2. If the platform gives us anything other than V1 or V2 then we shouldn't have gotten here or something is wrong with the `switch` statement in `CryptoSettings::writeFile()`.

File: src/platforms/linux/linuxcryptosettings.cpp
Comment (oskirby) [reply src/platforms/linux/linuxcryptosettings.cpp:66]: Sure, I can put the error message back.

File: taskcluster/kinds/build/linux.yml
Comment (mcleinman) [reply taskcluster/kinds/build/linux.yml:58]: Thanks. That makes sense. Do we need to ensure any help docs or download pages are updated w/ this info?

File: taskcluster/kinds/docker-image/kind.yml
Comment (mcleinman) [reply taskcluster/kinds/docker-image/kind.yml:60]: 👍🏻

File: src/platforms/linux/xdgportal.cpp
Comment (mcleinman) [reply src/platforms/linux/xdgportal.cpp:17]: good find

File: src/cryptosettings.cpp
Comment (mcleinman) [reply src/cryptosettings.cpp:335]: Thanks. Makes sense now.

**Review by mcleinman — APPROVED** [review]
I've added @lesleyjanenorton to this review - Like I said earlier, given that this is security-related and something I'm not that familiar with, I'd love a second r+ on this one. 

r+ from me, thanks for answering a bunch of questions as I get more familiar with this part of the code. I'm not marking "approve" yet so we can't merge it until we get a second reviewer who looks at it.

**Review by lesleyjanenorton — APPROVED** [review]
👍 On behalf of our futures selves, let's make sure to tell the folks in support about the launch quirk and maybe add something to the docs.

**Comment by oskirby** [comment]
After a bit of research, I think we could workaround the startup quirk by creating a Systemd scope on startup if we find that the application is launched outside of one. Basically we would need to invoke the [startTransientUnit()](https://www.freedesktop.org/software/systemd/man/latest/org.freedesktop.systemd1.html) method over D-Bus and request a scope name which matches the XDG desktop portal [parser implementation](https://github.com/flatpak/xdg-desktop-portal/blob/07284c093478afbb48305f3fe533ad5078a0085c/src/xdp-app-info-host.c#L80)

**Pull request merged**
